pkg/store/gate.go

```diff
@@ -0,0 +1,54 @@
+package store
+
+import (
+	"context"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/prometheus/pkg/gate"
+)
+
+// Gate wraps the Prometheus gate with extra metrics.
+type Gate struct {
+	g              *gate.Gate
+	currentQueries prometheus.Gauge
+	gateTiming     prometheus.Summary
+}
+
+// NewGate returns a new gate.
+func NewGate(maxConcurrent int, reg prometheus.Registerer) *Gate {
+	g := &Gate{
+		g: gate.New(maxConcurrent),
+	}
+	g.currentQueries = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "thanos_bucket_store_queries_total",
+		Help: "Total number of currently executing queries.",
+	})
+	g.gateTiming = prometheus.NewSummary(prometheus.SummaryOpts{
```

Comment:
Summaries should be banned. They are quite expensive and non aggregatable. Can we use histogram? (:

Reply:
Will do.